third_party/nvfuser/csrc/ir_internal_nodes.h

```diff
@@ -2181,4 +2229,121 @@ class TORCH_CUDA_CU_API NamedScalar : public Val {
   std::string name_;
 };
 
+class TORCH_CUDA_CU_API PadOp : public Expr {
+ public:
+  using Expr::Expr;
+
+  //! Pad a tensor as specified by a vector of integer scalars. For
+  //! the actual semantics, see the torch.pad documentation
+  PadOp(
+      IrBuilderPasskey passkey,
+      TensorView* out,
+      TensorView* inp,
+      const std::vector<Val*>& pad_widths);
+
+  NVFUSER_DECLARE_CLONE_AND_CREATE
+
+  virtual const char* getOpString() const override {
+    return "PadOp";
+  }
+
+  std::string toString(int indent_size = 0) const override;
+  std::string toInlineString(int indent_size = 0) const override;
+
+  Val* out() const {
+    return output(0);
+  }
+
+  Val* in() const {
+    return input(0);
+  }
+
+  std::vector<int> getPaddedAxes() const;
+
+  std::vector<Val*> getPadWidths() const;
+
+  std::pair<Val*, Val*> getPadWidths(int axis) const;
+};
+
+// Similar to at::indexing::Slice
+struct Slice {
+  Val* start = nullptr;
+  Val* stop = nullptr;
+  Val* step = nullptr;
+};
+
+class TORCH_CUDA_CU_API SliceOp : public Expr {
+ public:
+  using Expr::Expr;
+
+  SliceOp(
+      IrBuilderPasskey passkey,
+      TensorView* out,
+      TensorView* inp,
+      const std::vector<Slice>& ranges);
+
+  NVFUSER_DECLARE_CLONE_AND_CREATE
+
+  virtual const char* getOpString() const override {
+    return "SliceOp";
+  }
+
+  std::string toString(int indent_size = 0) const override;
+  std::string toInlineString(int indent_size = 0) const override;
+
+  Val* out() const {
+    return output(0);
+  }
+
+  Val* in() const {
+    return input(0);
+  }
+
+  std::vector<Slice> getRanges() const;
+};
+
+class TORCH_CUDA_CU_API CatOp : public Expr {
+ public:
+  using Expr::Expr;
+
+  CatOp(
+      IrBuilderPasskey passkey,
+      Val* out,
+      const std::vector<Val*>& inputs,
+      int concatenated_dim);
+
+  //! Create a cat op with the index and predicates for codegen. Only
+  //! used for the Kernel container
+  CatOp(
+      IrBuilderPasskey passkey,
+      Val* out,
+      const std::vector<Val*>& inputs,
+      int concatenated_dim,
+      Val* concatenated_domain_index,
+      const std::vector<Bool*>& preds);
+
+  NVFUSER_DECLARE_CLONE_AND_CREATE
+
+  virtual const char* getOpString() const override {
+    return "CatOp";
+  }
+
+  std::string toString(int indent_size = 0) const override;
+  std::string toInlineString(int indent_size = 0) const override;
+
+  int concatenatedDim() const {
+    return attribute(0)->as<Attribute<int>>()->value;
+  }
+
+  //! The index val that determines which input tensor should be used
+  //! to fill the particular output position of this expression. Only
+  //! valid after indexing
+  Val* getConcatenatedDomainIndex() const;
+
+  //! Gets a Bool indicating if the input tensor specified by
+  //! tensor_idx should be used to fill the output tensor. Only valid
+  //! with the Kernel container
+  Bool* getPred(int input_idx) const;
```

Comment:
Instead of putting lowering information here and handle it in `codegen.cpp`, does it make sense to lower cat to a sequence of ternary where op, or even crazier, `kir::IfThenElse`?

Reply:
Sorry, didn't see your comment in https://github.com/csarofeen/pytorch/pull/2480/files#diff-66e2e03e8b5e393d98fbcde0edfc1ef72ba81526bfaa8ef5a316907ebcff2892R1440-R1443

Reply:
Yeah, I think this is reasonable. We could extend `kir::IfThenElse` to have `else if` clauses as well, but that seems overkill if the only use case is here.